src/Renderer/c3DEngine.js

```diff
@@ -8,258 +8,256 @@ import * as THREE from 'three';
+            options.alpha = true;
+        }
+        if (options.logarithmicDepthBuffer === undefined) {
+            options.logarithmicDepthBuffer = this.gLDebug || NOIE;
+        }
 
-    if (!renderer && options.logarithmicDepthBuffer) {
-        // We don't support logarithmicDepthBuffer when EXT_frag_depth is missing.
-        // So recreated a renderer if needed.
-        if (!this.renderer.extensions.get('EXT_frag_depth')) {
-            const _canvas = this.renderer.domElement;
-            this.renderer.dispose();
-            this.renderer = new THREE.WebGLRenderer({
-                canvas: _canvas,
+        const renderer = rendererOrDiv.domElement ? rendererOrDiv : undefined;
+        const viewerDiv = renderer ? undefined : rendererOrDiv;
+
+        this.width = (renderer ? renderer.domElement : viewerDiv).clientWidth;
+        this.height = (renderer ? renderer.domElement : viewerDiv).clientHeight;
+
+        this.positionBuffer = null;
+        this._nextThreejsLayer = 1;
+
+        this.fullSizeRenderTarget = new THREE.WebGLRenderTarget(this.width, this.height);
+        this.fullSizeRenderTarget.texture.minFilter = THREE.LinearFilter;
+        this.fullSizeRenderTarget.texture.magFilter = THREE.NearestFilter;
+        this.fullSizeRenderTarget.depthBuffer = true;
+        this.fullSizeRenderTarget.depthTexture = new THREE.DepthTexture();
+        this.fullSizeRenderTarget.depthTexture.type = THREE.UnsignedShortType;
+
+        this.renderView = function _(view) {
+            this.renderer.clear();
+            this.renderer.render(view.scene, view.camera.camera3D);
+        }.bind(this);
+
+        this.onWindowResize = function _(w, h) {
+            this.width = w;
+            this.height = h;
+            this.fullSizeRenderTarget.setSize(this.width, this.height);
+            this.renderer.setSize(this.width, this.height);
+        }.bind(this);
+
+        // Create renderer
+        try {
+            this.renderer = renderer || new THREE.WebGLRenderer({
+                canvas: document.createElement('canvas'),
                 antialias: options.antialias,
                 alpha: options.alpha,
-                logarithmicDepthBuffer: false,
+                logarithmicDepthBuffer: options.logarithmicDepthBuffer,
             });
+        } catch (ex) {
+            console.error('Failed to create WebGLRenderer', ex);
+            this.renderer = null;
         }
-    }
 
-    // Let's allow our canvas to take focus
-    // The condition below looks weird, but it's correct: querying tabIndex
-    // returns -1 if not set, but we still need to explicitly set it to force
-    // the tabindex focus flag to true (see
-    // https://www.w3.org/TR/html5/editing.html#specially-focusable)
-    if (this.renderer.domElement.tabIndex === -1) {
-        this.renderer.domElement.tabIndex = -1;
-    }
+        if (!this.renderer) {
+            // from THREE/WebGL.js
+            const element = document.createElement('div');
```

Comment:
Could you move this html message error in specific file (with message in parameter)?

There's may be a redundancy with console error in line 63.

Reply:
I replaced it by calling the `WEBGL.getErrorMessage()` from `three` directly.